python/mlx/nn/layers/gnn/message_passing.py

```diff
@@ -0,0 +1,96 @@
+from typing import Any
+
+import mlx.core as mx
+import mlx.nn as nn
+
+
+class MessagePassing(nn.Module):
+    r"""Base class for creating Message Passing Neural Networks (MPNNs) [1].
+
+    Inherit this class to build arbitrary GNN models based on the message
+    passing paradigm. This implementation is inspired from PyTorch Geometric [2].
+
+    Args:
+        aggr (str): the aggregation strategy used to aggregate messages
+
+    References:
+        [1] Gilmer et al. Neural Message Passing for Quantum Chemistry.
+        https://proceedings.mlr.press/v70/gilmer17a/gilmer17a.pdf
+
+        [2] Fey et al. PyG
+        https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.nn.conv.MessagePassing.html
+    """
+
+    def __init__(self, aggr="add"):
+        super().__init__()
+
+        self.aggr = aggr
+
+    def __call__(self, x: mx.array, edge_index: mx.array, **kwargs: Any):
+        raise NotImplementedError
+
+    def propagate(self, x: mx.array, edge_index: mx.array, **kwargs: Any) -> mx.array:
```

Comment:
I know this is a draft, but it would be useful to support propagating arbitrary features over nodes or edges. A concrete example where this is useful is [TransformerConv](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/conv/transformer_conv.html), where we want to firstly compute ``q,k,v`` and then propagate them.

Reply:
Indeed, we have to see how to pass the args properly, without too much boilerplate code. I know pyg has a [lot of code](https://pytorch-geometric.readthedocs.io/en/latest/_modules/torch_geometric/nn/conv/message_passing.html#MessagePassing) dedicated to that with the various inspectors, which is not very clean.

Reply:
a quick fix could be to only leverage `(self, edge_index: mx.array, **kwargs: Any)` as args, and provide `x`, or `q, k, v` within `kwargs`.

Reply:
> a quick fix could be to only leverage `(self, edge_index: mx.array, **kwargs: Any)` as args, and provide `x`, or `q, k, v` within `kwargs`.

Sounds clean at first glance :)